core/src/main/java/com/datastax/oss/driver/internal/core/util/DefaultDependencyChecker.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A checker for the presence of various {@link Dependency} instances at runtime. Predicate tests
+ * for Graal substitutions should NOT use this class; see {@link GraalDependencyChecker} for more
+ * information.
+ */
+public class DefaultDependencyChecker {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultDependencyChecker.class);
+
+  private static ConcurrentHashMap<Dependency, Boolean> CACHE = new ConcurrentHashMap<>();
+
+  /**
+   * Return true iff we can find all classes for the dependency on the classpath, false otherwise
+   *
+   * @param dependency the dependency to search for
+   * @return true if the dependency is available, false otherwise
+   */
+  public static boolean isPresent(Dependency dependency) {
+    try {
+      return CACHE.computeIfAbsent(
+          dependency,
+          (dep) -> {
+            for (String classNameToTest : dependency.classes()) {
+              // Always use the driver class loader, assuming that the driver classes and
+              // the dependency classes are either being loaded by the same class loader,
+              // or – as in OSGi deployments – by two distinct, but compatible class loaders.
+              if (Reflection.loadClass(null, classNameToTest) == null) {
```

Comment:
Preserving comment from @adutra on the original PR:

> I wonder: can't we get rid of this class and use only the other one? My thinking is that calling
> Reflection.loadClass(null, classNameToTest) effectively boils down to Class.forName(classNameToTest) since the 
> CL is null.

This is a fairly significant point and I wanted to make sure it was discussed and addressed fully.

I definitely considered this, but in the end I thought it was worthwhile to keep this usage here.  My argument goes as follows: this checker is used in a number of runtime checks in normal non-Graal operations (the Graal case rules them out via substitutions).  In those cases I'd argue we _want_ to use the Reflection class; yes, the underlying mechanism resolves down to Class.forName() in that case as well but Reflection.loadClass() does a lot more logging about what's going on... and I could see that information being very useful when trying to debug runtime errors.

I _definitely_ see the argument for moving to a single implementation; believe me, I was very tempted.  But in the end I thought preserving the solid logging that is the status quo was more important.  I'm certainly open to argument if you still disagree, however.

Reply:
Hmm I think you are right here. Debugging class loading issues is sometimes tricky and having some logs to look at can definitely be a life saver. Let's keep the 2 impls.